wall_e/src/resources/cogs/ban.py

```diff
@@ -0,0 +1,419 @@
+                         ban_date=audit_ban.created_at.timestamp(),
+                         reason=audit_ban.reason if audit_ban.reason else 'No Reason Given!'
+                         )
+
+        # update ban_list and db
+        self.ban_list.append(member.id)
+        await BanRecords.insert_record(ban)
+
+        # unban
+        await self.bot.guilds[0].unban(member)
+        logger.info(f"[Ban intercept()] ban for {ban.username} moved into db and guild ban was removed")
+
+        # report to council
+        e_obj = discord.Embed(title="Ban Hammer Deployed",
+                              colour=discord.Color.red())
+
+        e_obj.add_field(name="Banned User", value=f"**{ban.username}**", inline=True)
+        e_obj.add_field(name="Moderator", value=f"**{ban.mod}**", inline=True)
+        e_obj.add_field(name="Reason", value=f"```{ban.reason}```", inline=False)
+        e_obj.add_field(name="Notification DM", value="*NOT SENT*\nCause: NO COMMON GUILD\n", inline=False)
+        e_obj.set_footer(text="Intercepted Moderator Action")
+        e_obj.timestamp = audit_ban.created_at
+        await self.mod_channel.send(embed=e_obj)
+        logger.info(f"[Ban ban()] Message sent to mod channel,{self.mod_channel}, of the ban for {ban.username}.")
+
+    @commands.command()
+    async def convertbans(self, ctx):
+        """Reads in all guild bans into this ban system"""
+
+        logger.info(f"[Ban convertbans()] convertbans command detected from {ctx.author}")
+
+        try:
+            # audit logs contains info about user who did the banning, the timestamp of the ban, and the reason
+            # however audit logs only go back 3 months, so have to read older bans from the bans list
+            ban_logs = await self.bot.guilds[0].audit_logs(action=discord.AuditLogAction.ban).flatten()
+            guild_ban_list = await self.bot.guilds[0].bans()
+        except Exception as e:
+            logger.info(f'[Ban convertbans()] error while fetching ban data: {e}')
+            await ctx.send(f"Encountered the following errors: {e}\n**Most likely need view audit log perms.**")
+            return
+
+        if not guild_ban_list:
+            logger.info("[Ban convertbans()] No bans to migrate into the ban system from guild. "
+                        "Sening message and ending command."
+                        )
+            await ctx.send("There are no bans to migrate from the guild to the wall_e ban systeme.")
+            return
+
+        logger.info(f"[Ban convertbans()] retrieved audit log data for ban actions: {ban_logs}")
+        logger.info(f"[Ban convertbans()] retrieved ban list from guild: {guild_ban_list}")
+
+        logger.info("[Ban convertbans()] Starting process to move all guild bans into db")
+
+        # update self.ban_list
+        self.ban_list.extend([ban.user.id for ban in guild_ban_list])
+        ban_records = []
+        ban_log_ids = []
+        for ban in ban_logs:
```

Comment:
is there a reason why you did not formulate the logic like this?

```
        try:
            # audit logs contains info about user who did the banning, the timestamp of the ban, and the reason
            # however audit logs only go back 3 months, so have to read older bans from the bans list
            ban_logs = {
                ban_log.target.id: ban_log
                for ban_log in await self.bot.guilds[0].audit_logs(action=discord.AuditLogAction.ban).flatten()
            }
            complete_guild_ban_list = await self.bot.guilds[0].bans()
        except Exception as e:
            logger.info(f'[Ban convertbans()] error while fetching ban data: {e}')
            await ctx.send(f"Encountered the following errors: {e}\n**Most likely need view audit log perms.**")
            return

        if not complete_guild_ban_list:
            logger.info("[Ban convertbans()] No bans to migrate into the ban system from guild. "
                        "Sening message and ending command."
                        )
            await ctx.send("There are no bans to migrate from the guild to the wall_e ban systeme.")
            return

        logger.info(f"[Ban convertbans()] retrieved audit log data for ban actions: {ban_logs}")
        logger.info(f"[Ban convertbans()] retrieved ban list from guild: {complete_guild_ban_list}")

        logger.info("[Ban convertbans()] Starting process to move all guild bans into db")

        # update self.ban_list
        ban_records = []
        for ban in complete_guild_ban_list:
            if ban.user.id not in self.ban_list:
                self.ban_list.append(ban.target.id)
                mod = None
                mod_id = None
                ban_date = None
                reason = 'No Reason Given!'
                if ban.user.id in ban_logs:
                    username = ban.target.name + '#' + ban.target.discriminator
                    user_id = ban.target.id
                    mod = ban.user.name + '#' + ban.user.discriminator
                    mod_id = ban.user.id
                    ban_date = ban.created_at.timestamp(),
                    reason = ban.reason if ban.reason else 'No Reason Given!'

                else:
                    username = ban.user.name + '#' + ban.user.discriminator
                    user_id = ban.user.id
                ban_records.append(BanRecords(
                    username=username,
                    user_id=user_id,
                    mod=mod,
                    mod_id=mod_id,
                    ban_date=ban_date,
                    reason=reason
                ))
```

Referring specifically to the fact that I have only 1 for loop while you have 2 which would save a lot of time if there are alot of bans.

Reply:
because you still need 2 loop no matter the case. also ur code wouldn't work:

* `self.ban_list.append(ban.target.id)`: here ban doesnt have a `target` property, only `user`
* `if ban.user.id in ban_logs` this'll always return false because `ban_logs` is a list auditLogEntries(glorified dicts)
  * you'd have to loop over every audit log to find a matching id(s). so for every ban in the ban list you'd run the entire audit log, since theres an edge case where someone could have more than 1 entry in the audit logs for a ban. this is very unlikely but a possibility 
  * if this somehow did return `True`, which it wouldn't. `ban` is still referring to a ban object which doesn't have the properties being access below like `created_at` or `reason`. 

so for `n` total guild bans and `m` ban logs my two loop  would have a total of `n + m` iterations while your single loop would have `n*m`. Given worst case `n = m` my loops run at `O(n)` while the single loop run at `O(n^2)`

Reply:
>self.ban_list.append(ban.target.id): here ban doesnt have a target property, only user

corrected

>if ban.user.id in ban_logs this'll always return false because ban_logs is a list auditLogEntries(glorified dicts) 

in my example, `ban_logs` is a dictionary where the key is `ban_log.target.id`

>because you still need 2 loop no matter the case. also ur code wouldn't work:

dict comprehensions' performance is better than a for loop.

Reply:
I added these changes. It actually runs the tiniest bit slower but the readability is better so worth imo.

Reply:
who da man? I da man 😎